text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+    /// time a future internal to that task is ready to make progress.
+    ///
+    /// The `poll` function is not called repeatedly in a tight loop for
+    /// futures, but only whenever the future itself is ready, as signaled via
+    /// `cx.waker()`. If you're familiar with the `poll(2)` or `select(2)`
+    /// syscalls on Unix it's worth noting that futures typically do *not*
+    /// suffer the same problems of "all wakeups must poll all events"; they
+    /// are more like `epoll(4)`.
+    ///
+    /// An implementation of `poll` should strive to return quickly, and should
+    /// *never* block. Returning quickly prevents unnecessarily clogging up
+    /// threads or event loops. If it is known ahead of time that a call to
+    /// `poll` may end up taking awhile, the work should be offloaded to a
+    /// thread pool (or something similar) to ensure that `poll` can return
+    /// quickly.
+    ///
+    /// # Panics
+    ///
+    /// Once a future has completed (returned `Ready` from `poll`), subsequent
+    /// calls to `poll` may panic, block forever, or otherwise cause bad behavior.
+    /// The `Future` trait itself provides no guarantees about the behavior of
+    /// `poll` after a future has completed.
+    fn poll(self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Item>;
+}
+```
+
+Most of the explanation here follows what we've already said about the task
+system.  The one new twist is the use of `Pin`, which makes it possible to keep
+data borrowed across separate calls to `poll` (i.e., "borrowing over yield
+points"). The mechanics of pinning are explained
+in [the RFC that introduced it](https://github.com/rust-lang/rfcs/pull/2349),
+but one key point is that when implementing `Future` by hand, if you don't wish
+to use this borrowing feature, you can treat `Pin<Self>` just like `&mut self`.
+
+### Universal adapters
+
+There are several basic adapters, modeled on `Iterator`, that work for arbitrary
+futures. Using `impl Trait` notation, we can define them as follows:
+
+```rust
+trait Future {
+    // Transform the result of the future
+    fn map<T>(self, f: impl FnOnce(Self::Item) -> T) -> impl Future<Item = T>
+        { .. }
+
+    // Chain a future onto this one
+    fn then<F>(self, f: impl FnOnce(Self::Item) -> F) -> impl Future<Item = F::Item>
+        where F: Future
+        { .. }
+
+    // Chain a closure for side effects
+    fn inspect(self, f: impl FnOnce(&Self::Item)) -> impl Future<Item = Self::Item>
+        { .. }
+
+    // Translate unwinding within this future into a `Result`
+    fn catch_unwind(self) -> impl Future<Item = Result<Self::Item, Box<Any + Send>>>
+        where Self: UnwindSafe
+        { .. }
+}
+```
+
+These adapters are all straightforward and based on substantial prior art.
+
+More interesting adapters like `shared`, `select` and `join` will be left to the
+ecosystem to iterate on before being RFC'ed for `std`.
+
+### `Result`-specific adapters
```

Comment:
Regarding `FutureRes<T, E>: Future<Item = Result<T, E>>`, has the impact of this change been checked against the ecosystem?

I ask because this is the exact route I experimented with a before the futures 0.2 RFC. I spiked out this exact pattern and experimented with updating existing code (not code like Tokio, but stuff more like tower, tower-grpc, h2) and the generic complexity quickly got out of hand.

The result of my experimentation was that I didn't propose it as I couldn't sanely update existing code.

I will try to dig out my past experiments or recreate them, but as a simple example:

// Today
impl<T: Future> Foo<T> { ... }

// Proposed
impl<T, E, F: FutureRes<T, E>> Foo<F> { ... }
```

You can see that what was a single generic now becomes three. This complexity balloons as the generics get more complicated.

```rust
// Today
impl<T: Future, U: Future> Foo<T, U> { ... }

// Proposed
impl<T1, T2, E1, E2, F1: FutureRes<T1, E1>, F2: FutureRes<T1, E2>> Foo<F1, F2> { ... }
```

That last scenario is a relatively simple scenario too and you can already see how it gets unwieldy. I can try to pull up the more complex examples that I encountered in Tower, Conduit, etc... there were some that I was unable to get compiling due to failing to match up the generics.

Reply:
I agree that the generics can get unwieldy-- it seems like what we really want is something like this:
```rust
trait FutureResult: Future<Output=Result<Self::Item, Self::Error>> {
    type Item;
    type Error;
    ...
}

impl<F, T, E> FutureResult for F where F: Future<Output=Result<T, E>> {
    type Item = T;
    type Error = E;
    ...
}
```
But that doesn't work because the `FutureResult` trait definition is cyclic. You really want to be able to write `F: FutureResult` and be able to project out the `Ok` and `Err` type variants (`T` and `E`).  We could do this by removing the `: Future<...>` bound on `FutureResult`, but then a `FutureResult` bound doesn't give you `FutureResult`. Maybe that's okay, though? Writing `F: Future + FutureResult` seems more ergonomic than the mess of generics you'd get from the current proposed solution. WDYT?

Reply:
@cramertj I can't really comment thoughtfully about `F: Future + FutureResult` without trying to actually use it.

That said, I don't see how your proposal actually works unless it includes assumptions that I am not aware of. Given `F: Future + FutureResult` there is nothing that says `<F as Future>::Item == Result<<F as FutureResult>::Item, <F as FutureResult>::Error>`.

Assuming it does work, you are still throwing out ergonomics for the 99% case to support less common cases.

Why not use `!` for `Error`  when the future is infallible? It is less "pure" of a solution, but afaik, 

* Not having an error is far less common. The only time that I have **personally** encountered that is for the type that `spawn` takes. Every other future had an error.
* `T: Future<Error = !>` still models the infallible case.

Reply:
(FWIW, GHC accepts the transliteration of that `trait` - but not the `impl`! It doesn't like the fact that the RHSs of the associated types mention type variables which weren't introduced on their LHSs. I think we have some kind of vaguely similar rule for unbound/unused/unconstrained(?) variables in `impl`s.)

Reply:
@cramertj 

I am sympathetic to the fact that there is no one great solution. I have also really wanted to remove the required `Error` associated type. 

As mentioned, I spent a lot of time experimenting w/ the option proposed in the RFC. I am pretty sure I touched on the route you suggested, but it didn't work due to the reason I mentioned.

My conclusion was that, given the current state of the rust compiler, it was not possible to remove the `Error` associated type while having a sane experience for network related usage of futures.

I really would be happy to be proven wrong because I agree that `Future<Error = !>` is hacky, but that hack is far less painful than the proposed alternatives.

Reply:
@cramertj To be clear I'm not sure if being able to accept the `trait` helps us at all unless we can *also* accept the `impl`. :) For which the fact that GHC doesn't doesn't bode well. (I don't have the expertise to know whether/why/how we could do differently - @nikomatsakis could maybe say.)

Reply:
I believe this could be made to work by extending the trait aliases RFC: https://github.com/rust-lang/rfcs/pull/1733

It could look something like:
```rust
trait FutureRes { type Item; type Error; } = Future<Output=Result<Item, Error>>;
```

Alternatively, you could do it with associated type defaults:
```rust
trait ResultHelper {
    type Item;
    type Error;
}

impl<T, E> ResultHelper for Result<T, E> {
    type Item = T;
    type Error = E;
}

trait Future {
    ...
    type Item where Self::Output: ResultHelper = Self::Output::Item;
    type Error where Self::Output: ResultHelper = Self::Output::Error;
}

trait FutureRes = Future where Self::Output: ResultHelper;
```

Reply:
I really appreciate the "purity" of not tying the output to a `Result`, but to just *something*. But, we definitely can't accept making things much worse for those using futures in fallible contexts (like networking). But, maybe with a secondary trait (of Future) is needed.

**We have `Try`!** It's got associated types for ok and err, is already in core, and implemented for `Result`. And it could even allow uses that aren't `Result`...

```rust
trait Future {
    type Item;

    // But that name... XD
    fn map_ok<F, U>(self, f)
    where
        Self::Item: Try,
        F: FnOnce(Self::Item::Ok) -> U;
}
```

I haven't tried this myself, since the trait is unstable. Would it stabilize in time? Looking at the tracking issue, they seem to be discussing some peculiar cases, removing the error-ness of the trait...

Reply:
@seanmonstar Interesting. I have not experimented with this exact route (I did experiment w/ `where Self: Future<Item = Result<T, U>>`)

As a quick experiment:

```rust
// Today
fn foo<T>(f: F)
where T: Future<Error = io::Error>
{ ... }

// `Self::Item: Try` bound
fn foo<T>(f: F)
where T: Future,
           T::Item: Try<Error = io::Error>,
{ ... }
```

Not great, but better than the equivalent with `FutureResult`.

To consider, (as far as I know), I/O will be the first entry point for the majority of users and adds non trivial complexity to do what used to be simple.

Reply:
@cramertj 

> In my experience, however, the balance shifts significantly in favor of infallible futures when writing application code as opposed to libraries. In applications, it's common to handle errors immediately by logging or doing some other action, rather than propagating the error up the stack.

This does not match my experience working on applications w/ Futures. Errors tend to be converted to log messages or otherwise handled at the very root of the future tree (the spawn point).

However, wouldn't async / await basically handle the application cases? At that point, you would probably just:

```rust
if let Err(e) = await!(some_op()) {
    error!("an error happened");
    return;
}
```

In that case, whether the future is represented as an infallible trait or `Future<Error = !>` is not visible.

The bigger ongoing issue is going to be setting trait bounds on generics.

Reply:
@carllerche When writing an `async fn` that handles errors the way you suggested (`if let Err(e) ... { log }`), what would you expect to put as the return type of that function? I'd expect to be able to write `async fn foo() { ... }`, not `async fn foo() -> Result<(), !> { ... }`.

Reply:
@cramertj couldn't the `async` expansion check the return value, and if it is `Result`, leave it alone, otherwise wrap it as `Result<Ret, !>`?

Reply:
@seanmonstar Yeah, that's what I meant when I said
> We could implicitly convert non-Result types into Result<T, !>, but then presumably users would still have to match on the Result that came from awaiting or polling the result.

e.g.

```rust
async fn bar() -> i32 { 5 }
fn foo() -> i32 {
    5 + block_on(bar()) // ERROR: cannot add `{number}` and `Result`
}
```

Reply:
If you stick an `unwrap` in there (which should be a no-op because of `!`), it'll work. Or, eventually you could destructure the `Ok`:

```rust
async fn bar() -> i32 { 5 }
fn foo() -> i32 {
    let Ok(b) = block_on(bar());
    5 + b
}
```

Reply:
@cramertj My response is [here](https://github.com/rust-lang/rfcs/pull/2395#issuecomment-380611871) (too big to make as an inline comment).